driver/src/orderbook/streamed/balance.rs

```diff
@@ -2,63 +2,104 @@ use super::*;
-            .balance
+    pub fn withdraw(&mut self, amount: U256, batch_id: BatchId) -> Result<(), Error> {
+        if self.withdraw.batch_id >= batch_id {
+            return Err(Error::WithdrawEarlierThanRequested(self.withdraw.batch_id));
+        }
+        if self.withdraw.amount < amount {
+            return Err(Error::WithdrawMoreThanRequested(self.withdraw.amount));
+        }
+        let balance_excluding_withdraw = self.get_balance_internal(batch_id, false)?;
+        let new_balance = balance_excluding_withdraw
             .checked_sub(amount)
-            .ok_or(Error::MathUnderflow)?;
+            .ok_or(Error::WithdrawMoreThanBalance(balance_excluding_withdraw))?;
+        self.balance = new_balance;
+        // If there is a valid deposit then new_balance already includes it so we need to clear it
+        // now.
+        self.clear_pending_deposit_if_possible(batch_id);
+        // Works like in the smart contract: Any withdraw unconditionally removes the withdraw
+        // request even if the amount is smaller than requested.
+        self.withdraw.amount = 0.into();
         Ok(())
     }
 
-    pub fn update_deposit_balance(&mut self, current_batch_id: BatchId) {
-        match self.pending_deposit {
-            Some(ref deposit) if deposit.batch_id < current_batch_id => {
-                self.balance += deposit.amount;
-                self.pending_deposit = None;
-            }
-            _ => (),
-        };
+    pub fn get_balance(&self, batch_id: BatchId) -> Result<U256, Error> {
+        self.get_balance_internal(batch_id, true)
     }
 
-    pub fn current_balance(&self, current_batch_id: BatchId) -> U256 {
+    fn get_balance_internal(
+        &self,
+        batch_id: BatchId,
+        include_withdraw: bool,
+    ) -> Result<U256, Error> {
         let mut balance = self.balance;
-        if let Some(ref flux) = self.pending_deposit {
-            if flux.batch_id < current_batch_id {
-                balance = balance.saturating_add(flux.amount);
-            }
+        if self.deposit.batch_id < batch_id {
+            balance = balance
+                .checked_add(self.deposit.amount)
+                .ok_or(Error::MathOverflow)?;
+        }
+        if include_withdraw && self.withdraw.batch_id < batch_id {
+            // Saturating because withdraw requests can be for amounts larger than balance.
+            balance = balance.saturating_sub(self.withdraw.amount);
         }
-        if let Some(ref flux) = self.pending_withdraw {
-            if flux.batch_id < current_batch_id {
-                balance = balance.saturating_sub(flux.amount);
-            }
+        Ok(balance)
+    }
+
+    fn apply_existing_deposit(&mut self, current_batch_id: BatchId) -> Result<(), Error> {
+        if self.deposit.batch_id < current_batch_id {
+            let new_balance = self
+                .balance
+                .checked_add(self.deposit.amount)
+                .ok_or(Error::MathOverflow)?;
```

Comment:
Maybe we can add backtraces to some of these errors? I feel like if we get a `MathOverflow` error in the logs it might be really hard to figure out where its coming from.

Reply:
I could add more variants to all the places where an overflow can occur but it feels like overkill and premature. If we get an error on the real events we will know at which events it happens and can replay that to debug.
What exactly do you mean by backtrace, how do I get one? Maybe I could use the line number to help us debug.

Reply:
If you use [`thiserror`](https://docs.rs/thiserror/1.0.15/thiserror/) you can add a `std::backtrace::Backtrace` field to your variants and the generated code takes care of the rest. It will include line number and the complete call stack in the error.

Reply:
This makes it so you need less variants since the actual call hierarchy gets captured by the event. It's basically the backtrace you see when you panic with `RUST_BACKTRACE=1`.

Reply:
I would like getting an error to be inexpensive. Getting a backtrace might be expesnive and it is going to be heap allocated. On the other hand errors are still rare and I haven't measured performance. Maybe we can keep this in mind if it turns out to would have been useful in the future.

Reply:
Errors should be exceptional cases, I don't think there is value optimizing the error path in our case. Also expensive is a relative term. In this case the cost is tiny compared to the time and memory it takes to perform the web3 request.

I would rather include this right away instead to make this section of code more debuggable. Also, `anyhow` creates backtraces on all errors, so its not like we don't already have that cost in other places.

Reply:
I mean `State` specifically because it will have a lot of events flowing through it. It still just doesn't feel right to include the backtrace in an error enum but it is not a strong opinion of mine. So if @fleupold also prefers it I'll add it.

Reply:
The backtrace is boxed so it only takes `usize` size in the error and only gets allocated on error. This is nice since errors are exceptional cases and so this will, most of the time, have 0 cost.

Reply:
When `anyhow` wraps an error it either reuses the backtrace of the error if there is one, or creates its own. So I think in either case a backtrace will be created, if it is part of the `MathOverflow` error for example, then it will just be more correct.